src/nearest_neighbors/datasets/heartsteps/loader.py

```diff
@@ -0,0 +1,338 @@
+
+
+    def process_data_scalar(self, cached:bool=False, agg:str="mean") -> tuple[np.ndarray, np.ndarray]:
+        """Process the data into scalar setting."""
+        df_steps, df_suggestions = self._load_data(cached)
+        Data, _, Mask = self._proc_dist_data(df_steps, df_suggestions)
+        if agg == "mean":
+            Data = Data.mean(axis=2)
+        elif agg == "sum":
+            Data = Data.sum(axis=2)
+        elif agg == "median":
+            Data = np.median(Data, axis=2)
+        elif agg == "std":
+            Data = np.std(Data, axis=2)
+        elif agg == "variance":
+            Data = np.var(Data, axis=2)
+        else:
+            raise ValueError(
+                "agg must be one of 'mean', 'sum', 'median', 'std', or 'variance'"
+            )
+
+        Data = np.squeeze(Data)
+        Data = Data.astype(object)
+        # write data to output_dir
+        if self.save_processed != "":
+            np.save(os.path.join(self.save_dir, "data.npy"), Data, allow_pickle=True)
+            np.save(os.path.join(self.save_dir, "mask.npy"), Mask, allow_pickle=True)
+        print("Done!")
+        return Data, Mask
+
+
+    def process_data_distribution(self, cached:bool=False) -> tuple[np.ndarray, np.ndarray]:
+        """Process the data into distribution setting."""
+        df_steps, df_suggestions = self._load_data(cached)
+
+        _, Data2d, Mask = self._proc_dist_data(df_steps, df_suggestions)
+        print("Done!")
+        return Data2d, Mask
+
+    ## HELPER FUNCTIONS
+    def _load_data(self, cached:bool=False) -> tuple[pd.DataFrame, pd.DataFrame]:
+        """Load the data from the remote source through urls."""
+        if cached:
+            jp_path = os.path.join(self.save_dir, "jbsteps.csv") 
+            sug_path = os.path.join(self.save_dir, "suggestions.csv")
+            if not (os.path.exists(jp_path) and os.path.exists(sug_path)):
+                print("No cached data found. Retrieving from url...")
+                jp_path = self.urls["jbsteps.csv"]
+                sug_path = self.urls["suggestions.csv"]
+        else:
+            jp_path = self.urls["jbsteps.csv"]
+            sug_path = self.urls["suggestions.csv"]
+       
+        df_steps = pd.read_csv(jp_path, low_memory=False)
+        df_suggestions = pd.read_csv(sug_path, low_memory=False)
+        return df_steps, df_suggestions
+
+    def _transform_dnn(
+        self,
+        df : Any,
+        users: int = 37,
+        max_study_day: int = 52,
+        day_dec: int = 5,
+        num_measurements: int = 12,
+    ) -> tuple[np.ndarray, np.ndarray]:
+        """Iteratively transform the processed HeartSteps data into a 4d tensor"""
+        warnings.simplefilter(action='ignore', category=pd.errors.PerformanceWarning)
+        final_M = np.zeros((users, max_study_day, day_dec, num_measurements))
+        final_A = np.zeros((users, max_study_day, day_dec))
+
+        for user in range(1, users + 1):
```

Comment:
This whole triple nested loop seems... a bit much? @albertgong1, do you have any feedback on this section?